### DISCUSSION TITLE
Problem in running the example zundel

### DISCUSSION THREAD

**Discussion by wangjiuning123** [MarsalekGroup/aml#26] [Q&A] [answered]
When I change input the element O to C for test ,CP2K is ok to run.
But in the second QbC step ,it is reported ''ValueError: At least some values of "forces" were undefined (`None`) for label "pred*"." 
What's  wrong ?

**Comment by beckobert** [comment, marked as answer]
The label property "pred*" refers to properties calculated during prediction, therefore the problem likely came up during the prediction of either training or remaining candidate structures with N2P2.
In this case the prediction failed because the structures in the zundel example contain only hydrogen and oxygen atoms. When you replace oxygen with carbon in your input file without changing the structures, N2P2 will fail because it can't find any carbon (and doesn't know what to do with the oxygen). I'm actually a bit surprised that the error didn't already occur earlier during training. If you want to make such a change, you will have to make the changes consistently throughout the example, which means you have to run MD with a suitable system that contains carbon.

Furthermore: For each iteration the calculations are saved in directories named `iteration-*` and you can look there for the log files, which usually give you a more detailed report of what went wrong and why.

**Reply by wangjiuning123** [reply]
Thanks for the reply.Actually, I mean I changed O to C in the AIMD run in the zundel example. AIMD run is OK and I get the data. There is my input structure:
4
water with unit cell: ABC [angstrom] 12.42 12.42 12.42
  C         3.8585873763       -4.7533175213        5.5091974759
  H         4.2109950951       -5.5568705259        5.9994585948
  C         3.0947084560       -4.3172663108        5.9135950646
  H        -3.5460761891        1.7456237131        4.2880212708

Here is my model:
kwargs_model = dict(
    elements = ('C','H'),
    n = 8,
    fn_template = 'input.nn',
    n_tasks = 8,
    n_core_task = 3,
    remove_output = True
)

But in the second QbC step,it is raise error like blow:
Prediction for training set... done.
Prediction for candidate set... done.
Traceback (most recent call last):
  File "./run-QbC.py", line 66, in <module>
    qbc.run()
  File "/opt/apps/apps/aml/aml/qbc.py", line 151, in run
    cda = evaluate_cda_forces(structures_candidate, 'pred*')
  File "/opt/apps/apps/aml/aml/qbc.py", line 28, in evaluate_cda_forces
    cda = structures.reduce_property(f_reduce=mean_std, name_prop='forces', label_prop=label_prop).mean()
  File "/opt/apps/apps/aml/aml/structures.py", line 383, in reduce_property
    data = s.properties.reduce_property(f_reduce=f_reduce, name_prop=name_prop, label_prop=label_prop)
  File "/opt/apps/apps/aml/aml/structures.py", line 650, in reduce_property
    raise ValueError(msg)
ValueError: At least some values of "forces" were undefined (`None`) for label "pred*".

**Reply by beckobert** [reply]
Oh, I misunderstood your original problem.

Based on this error message I can't tell what the problem is. I suggest two possible further steps.
1. Having looked at your input structure in vmd I wonder if the AIMD leads to any reasonable results, it looks very odd. In particular, the 2nd hydrogen is far away from the other atoms. So, if N2P2 does not find any other atom within the cutoff radius specified in your `input.nn` file for any atom (because one atom is isolated), prediction will likely fail.
2. For debugging I recommend to switch `remove_output` to `False` and rerun the QbC. That way you can inspect at the logs created by N2P2, which are usually more informative.

Let me know if any of this helps.

**Reply by wangjiuning123** [reply]
Thanks.I tried, It is seem to be the structure problem causing  I make up the structure randomly. Even though, I think maybe you can add this  
situation. May be An isolated atom in the cutoff radius can happens? The force could be assign to zero?

**Reply by beckobert** [reply]
Unfortunately, this issue would need to be addressed by the developers of N2P2. AML only processes the output of N2P2.
However, even for them, it might not be possible to address it, as most current MLPs are not suited to deal with this problem. They make their prediction based on the information of the local environment of each atom. If the local environment is empty, then there is nothing to base the prediction on. So it can either predict total garbage or throw an error.
